File: pyCombinatorial/algorithm/s_gui.py
############################################################################

# Created by: Prof. Valdecy Pereira, D.Sc.
# UFF - Universidade Federal Fluminense (Brazil)
# email:  valdecy.pereira@gmail.com
# Lesson: Guided Search
 
# GitHub Repository: <https://github.com/Valdecy>

############################################################################

# Required Libraries
import random
import copy

############################################################################

# Function: Tour Distance
def distance_calc(distance_matrix, city_tour):
    distance = 0
    for k in range(0, len(city_tour[0])-1):
        m        = k + 1
        distance = distance + distance_matrix[city_tour[0][k]-1, city_tour[0][m]-1]            
    return distance

# Function: Stochastic 2_opt
def stochastic_2_opt(distance_matrix, city_tour):
    best_route = copy.deepcopy(city_tour)      
    i, j       = random.sample(range(0, len(city_tour[0])-1), 2)
    if (i > j):
        i, j = j, i
    best_route[0][i:j+1] = list(reversed(best_route[0][i:j+1]))           
    best_route[0][-1]    = best_route[0][0]              
    best_route[1]        = distance_calc(distance_matrix, best_route)                     
    return best_route

# Function: Local Search
def local_search(distance_matrix, city_tour, penalty, max_attempts = 50, limit= 1):
    count    = 0
    ag_cost  = augumented_cost(distance_matrix, city_tour, penalty, limit)
    solution = copy.deepcopy(city_tour) 
    while (count < max_attempts):
        candidate           = stochastic_2_opt(distance_matrix, solution)
        candidate_augmented = augumented_cost(distance_matrix, candidate, penalty, limit)       
        if candidate_augmented < ag_cost:
            solution  = copy.deepcopy(candidate)
            ag_cost   = augumented_cost(distance_matrix, solution, penalty, limit)
            count     = 0
        else:
            count = count + 1                             
    return solution 

############################################################################

#Function: Augmented Cost
def augumented_cost(distance_matrix, city_tour, penalty, limit):
    augmented = 0   
    for i in range(0, len(city_tour[0]) - 1):
        c1 = city_tour[0][i]
        c2 = city_tour[0][i + 1]      
        if c2 < c1:
            c1, c2 = c2, c1            
        augmented = augmented + distance_matrix[c1-1, c2-1] + (limit * penalty[c1-1][c2-1])    
    return augmented

#Function: Utility
def utility (distance_matrix, city_tour, penalty, limit = 1):
    utilities = [0 for i in city_tour[0]]
    for i in range(0, len(city_tour[0]) - 1):
        c1 = city_tour[0][i]
        c2 = city_tour[0][i + 1]      
        if c2 < c1:
            c1, c2 = c2, c1            
        utilities[i] = distance_matrix[c1-1, c2-1] /(1 + penalty[c1-1][c2-1])  
    return utilities

#Function: Update Penalty
def update_penalty(penalty, city_tour, utilities):
    max_utility = max(utilities)   
    for i in range(0, len(city_tour[0]) - 1):
        c1 = city_tour[0][i]
        c2 = city_tour[0][i + 1]         
        if c2 < c1:
            c1, c2 = c2, c1        
        if (utilities[i] == max_utility):
            penalty[c1-1][c2-1] = penalty[c1-1][c2-1] + 1   
    return penalty

############################################################################

# Function: Guided Search
def guided_search(distance_matrix, city_tour, alpha = 0.3, local_search_optima = 1000, max_attempts = 20, iterations = 50, verbose = True):
    count         = 0
    limit         = alpha * (local_search_optima / len(city_tour[0]))  
    penalty       = [[0 for i in city_tour[0]] for j in city_tour[0]]
    solution      = copy.deepcopy(city_tour)
    best_solution = [[],float('+inf')]
    while (count < iterations):
        if (verbose == True):
            print('Iteration = ', count, 'Distance = ', round(best_solution[1], 2)) 
        solution  = local_search(distance_matrix, city_tour = solution, penalty = penalty, max_attempts = max_attempts, limit = limit)
        utilities = utility(distance_matrix, city_tour = solution, penalty = penalty, limit = limit)
        penalty   = update_penalty(penalty = penalty, city_tour = solution, utilities = utilities)
        if (solution[1] < best_solution[1]):
            best_solution = copy.deepcopy(solution) 
        count = count + 1
    route, distance = best_solution
    return route, distance

############################################################################

File: pyCombinatorial/web/js/algorithms/s_gui.js
// ============================================================================
// s_gui.js — Guided (Local) Search
// Mirrors algorithm/s_gui.py — augmented cost = real cost + λ × penalty;
// penalize utility-maximizing edges between iterations.
// ============================================================================

import { distanceCalc } from '../core/distance.js';
import { seedFunction, makeRng } from '../core/seed.js';

export const meta = {
  id: 's_gui',
  label: 'Guided Search',
  category: 'Local search',
  description: 'Penalize "bad" edges via augmented cost; iterate stochastic 2-opt.',
  params: [
    { key: 'iterations',  label: 'Iterations',     type: 'int',   default: 50,   min: 1,   max: 1000 },
    { key: 'maxAttempts', label: 'LS attempts',    type: 'int',   default: 20,   min: 1,   max: 1000 },
    { key: 'alpha',       label: 'α (penalty)',    type: 'float', default: 0.3,  min: 0,   max: 5, step: 0.05 },
    { key: 'optimaHint',  label: 'Optimum hint',   type: 'float', default: 1000, min: 0,   max: 1e7,
      hint: 'Reference local-optimum scale; tune to your instance.' },
    { key: 'seed',        label: 'Random seed',    type: 'int',   default: 42,   min: 0,   max: 99999 },
  ],
};

function key(c1, c2) { if (c2 < c1) [c1, c2] = [c2, c1]; return c1 * 1e6 + c2; }

function augCost(distanceMatrix, tour, penalty, limit) {
  let s = 0;
  for (let i = 0; i < tour.length - 1; i++) {
    let c1 = tour[i] - 1, c2 = tour[i + 1] - 1;
    if (c2 < c1) [c1, c2] = [c2, c1];
    s += distanceMatrix[c1][c2] + limit * (penalty.get(key(c1, c2)) || 0);
  }
  return s;
}

function stoch2opt(distanceMatrix, tour, rng) {
  const n = tour.length - 1;
  let i = Math.floor(rng() * n), j = Math.floor(rng() * n);
  if (i === j) j = (j + 1) % n;
  if (i > j) [i, j] = [j, i];
  const cand = tour.slice();
  const seg = cand.slice(i, j + 1).reverse();
  for (let k = 0; k < seg.length; k++) cand[i + k] = seg[k];
  cand[cand.length - 1] = cand[0];
  return [cand, distanceCalc(distanceMatrix, cand)];
}

function gLocalSearch(distanceMatrix, tour, dist, penalty, limit, maxAttempts, rng) {
  let solution = [tour.slice(), dist];
  let agCost = augCost(distanceMatrix, solution[0], penalty, limit);
  let count = 0;
  while (count < maxAttempts) {
    const c = stoch2opt(distanceMatrix, solution[0], rng);
    const ag = augCost(distanceMatrix, c[0], penalty, limit);
    if (ag < agCost) { solution = c; agCost = ag; count = 0; }
    else count++;
  }
  return solution;
}

export async function* run(distanceMatrix, params) {
  const rng = makeRng(params.seed ?? 42);
  let solution = params._initialTour
    ? [params._initialTour.slice(), distanceCalc(distanceMatrix, params._initialTour)]
    : seedFunction(distanceMatrix, rng);
  let best = [solution[0].slice(), solution[1]];

  const limit = (params.alpha ?? 0.3) * ((params.optimaHint ?? best[1]) / solution[0].length);
  const penalty = new Map();

  yield {
    phase: 'init', tour: best[0], bestTour: best[0],
    distance: best[1], bestDistance: best[1],
    iteration: 0, message: `λ=${limit.toFixed(3)}, start ${best[1].toFixed(2)}`,
  };

  const iterations = params.iterations ?? 50;
  const maxAttempts = params.maxAttempts ?? 20;

  for (let it = 0; it < iterations; it++) {
    solution = gLocalSearch(distanceMatrix, solution[0], solution[1], penalty, limit, maxAttempts, rng);
    // utilities; penalize max-utility edges
    const us = new Array(solution[0].length).fill(0);
    let mx = -Infinity;
    for (let i = 0; i < solution[0].length - 1; i++) {
      let c1 = solution[0][i] - 1, c2 = solution[0][i + 1] - 1;
      if (c2 < c1) [c1, c2] = [c2, c1];
      const p = penalty.get(key(c1, c2)) || 0;
      us[i] = distanceMatrix[c1][c2] / (1 + p);
      if (us[i] > mx) mx = us[i];
    }
    for (let i = 0; i < solution[0].length - 1; i++) {
      if (us[i] === mx) {
        let c1 = solution[0][i] - 1, c2 = solution[0][i + 1] - 1;
        if (c2 < c1) [c1, c2] = [c2, c1];
        penalty.set(key(c1, c2), (penalty.get(key(c1, c2)) || 0) + 1);
      }
    }
    const improved = solution[1] < best[1];
    if (improved) best = [solution[0].slice(), solution[1]];
    yield {
      phase: improved ? 'improvement' : 'penalty-update',
      tour: solution[0], bestTour: best[0],
      distance: solution[1], bestDistance: best[1],
      iteration: it + 1,
      message: improved ? `★ ${solution[1].toFixed(2)}` : `${solution[1].toFixed(2)}  best ${best[1].toFixed(2)}`,
    };
  }
  return { tour: best[0], distance: best[1], summary: `Guided Search: ${best[1].toFixed(2)}` };
}
